src/workerd/api/streams/standard.c++

```diff
@@ -1539,10 +1552,23 @@ void ReadableStreamBYOBRequest::respondWithNewView(jsg::Lock& js, jsg::BufferSou
                  TypeError,
                  "The view byte length must be zero after the stream is closed.");
   } else {
+    JSG_REQUIRE(!impl.controller->isExpectingClose(),
+                TypeError,
+                "Cannot respond with more bytes when expecting a close.");
     JSG_REQUIRE(view.size() > 0,
                  TypeError,
                  "The view byte length must be more than zero while the stream is open.");
+
+    size_t atLeast = impl.readRequest->getAtLeast();
+
     impl.readRequest->respondWithNewView(js, kj::mv(view));
+
+    // If the bytesWritten are less than atLeast, then this should be the final chunk
+    // of data we process. The next call to pull should close or error the ReadableStream.
+    if (view.size() < atLeast) {
+      impl.controller->expectCloseOnNextPull();
+    }
+
     pull = true;
```

Comment:
Nit: this line would make more sense right after `impl.readRequest->respondWithNewView()` for symmetry with `respond()` (or vice versa).